Add enable checkboxes to editor sections

lodetrick opened the pull request:
Supersedes #81630
Implements: https://github.com/godotengine/godot-proposals/issues/4173

Creates support for checkable property groups through the `PROPERTY_HINT_GROUP_ENABLE` property hint that is given to the `enabled` boolean. This property hint modifies the (sub)group to display a checkbox on its right. When the checkbox is clicked, the linked property is enabled and the group is unfolded. This PR also hides the corresponding "enable" property when it is inside a checkable group, saving vertical screen space.

This PR also makes the change that when a property group does not have any children, then it is unable to be unfolded and does not draw the unfolding arrow. This can be seen in the `Rotation Smoothing` property group in the example images.

| Before (4.4, `Limit` didn't have an enable option yet)  | After |
| ------------- | ------------- |
| <img width="280" alt="Before" src="https://github.com/user-attachments/assets/24bd0e28-8b22-4302-acd2-9238d9d4ca4e" />  | <img width="275" alt="After" src="https://github.com/user-attachments/assets/eb6ddcc8-1c1d-4519-b98c-df7ecd4264f3" />  |

As opposed to #81630, this PR chooses to have the checkboxes align to the right. This change makes the position of the text consistent, as before the text would jump to the right when the group was checkable.

cier committed:
Add enable checkboxes to editor sections

Fixes https://github.com/godotengine/godot-proposals/issues/4173

KoBeWi commented:
Disabled properties can still be clicked and it does nothing:

https://github.com/user-attachments/assets/25dbcc95-d5b7-4f0f-9a22-7f8a2eef38e0

It shouldn't display clicked StyleBox.

KoBeWi commented on editor/editor_inspector.cpp:
```diff
@@ -1856,10 +1862,34 @@ void EditorInspectorSection::_notification(int p_what) {
+						checkbox_position.x = get_size().width - (margin_end + checkbox->get_width());
+					}
+
+					Color color2(1, 1, 1);
```
```suggestion
					Color checkbox_color(1, 1, 1);
```

KoBeWi commented on editor/editor_inspector.cpp:
```diff
@@ -1856,10 +1862,34 @@ void EditorInspectorSection::_notification(int p_what) {
+						color2.g *= 1.2;
+						color2.b *= 1.2;
+					}
+					checkbox_position.y = (header_height - checkbox->get_height()) / 2;
```
This should be moved a above, so it's together with x.

KoBeWi commented on editor/editor_inspector.cpp:
```diff
@@ -1972,6 +2002,53 @@ Size2 EditorInspectorSection::get_minimum_size() const {
+	if (object->has_method("_get_property_warning")) {
+		const String custom_warning = object->call("_get_property_warning", related_enable_property);
+		if (!custom_warning.is_empty()) {
+			prologue = "[b][color=" + get_theme_color(SNAME("warning_color")).to_html(false) + "]" + custom_warning + "[/color][/b]";
```
The group color should be yellow too:
![image](https://github.com/user-attachments/assets/3a39ee03-5c54-4d7a-86de-9144fd8d5709)
Also looks like the warning color does not work here for some reason.

Compare with regular property:
![image](https://github.com/user-attachments/assets/f942665f-6d9b-4778-9009-9230cdde6089)

KoBeWi commented on editor/editor_inspector.cpp:
```diff
@@ -3930,6 +4046,43 @@ void EditorInspector::update_tree() {
+						last_created_section->set_checkable(p.name);
+						last_created_section->set_checked(value_checked.operator bool());
+
+						if (use_doc_hints) {
+							// `|` separators used in `EditorHelpBit`.
+							if (theme_item_name.is_empty()) {
+								if (p.name.contains("shader_parameter/")) {
+									ShaderMaterial *shader_material = Object::cast_to<ShaderMaterial>(object);
+									if (shader_material) {
+										last_created_section->set_tooltip_text("property|" + shader_material->get_shader()->get_path() + "|" + property_prefix + p.name);
+									}
+								} else if (p.usage & PROPERTY_USAGE_INTERNAL) {
+									last_created_section->set_tooltip_text("internal_property|" + classname + "|" + property_prefix + p.name);
+								} else {
+									last_created_section->set_tooltip_text("property|" + classname + "|" + property_prefix + p.name);
+								}
+							} else {
+								last_created_section->set_tooltip_text("theme_item|" + classname + "|" + theme_item_name);
+							}
```
Duplicated from EditorProperty.
It would be nice to remove the duplication (likely by adding a method), but it doesn't need to be done in this PR.

lodetrick replied:
The way the tooltip text is generated is not tied to the editor, so I moved it above and stored it in a variable that both instances use.

KoBeWi commented:
Looks like the feature is broken after your last push.

lodetrick commented:
> Looks like the feature is broken after your last push. Sections don't have checkboxes.

It should be fixed now. The comments on `vbox_added` implied that there was an optimization not to add the vbox until it was needed (like in an unfold), so I removed the `_test_unfold` when `foldable` is true. This caused the error, so I reverted it to how master always calls the method. I don't know then what optimization it is referring to, if it is breaking to add the vbox late. There might be an issue somewhere here

KoBeWi commented:
So uh https://github.com/godotengine/godot/pull/105272#issuecomment-2802010895
I only mentioned clicking, but you also removed hovering. Hover style is still relevant because of tooltips.

KoBeWi reviewed (approved):
Looks good now.

The is nice, but there are some potential concerns:
- The property name is hidden behind tooltip. Usually it's Section Title + Enabled, so it's intuitive.
- Enabling the property does not count it as changed (for the changes label). Either way can be considered correct though.
- The checkbox is more difficult to click than when it was a property. Maybe it could include some label, like "Enabled"?

I think these can be improved in the future, based on the feedback; the current state is not bad.

lodetrick commented:
One concern is that the section currently needs to display both the section title, the changes text (e.g. `(2 changes)`) when folded, the folding arrow, and now a checkbox. The horizontal space for the section is at a premium, so there would probably have to be a priority of which features to display if we add more text. It might be worth it to add a small `On` label like in the `EditorPropertyCheck`.

arkology commented:
> Since the enable properties are hidden when in a checkable group, this also makes their tooltips unreachable.

> One concern is that the section currently needs to display both the section title, the changes text (e.g. (2 changes)) when folded, the folding arrow, and now a checkbox.

There is another one concern: 
While property is disabled, it is impossible to see and edit corresponding properties like camera limits, sprite region ect

lodetrick commented:
This PR does not change the visibility of any property other than the `enabled` property (which it moves into the section header). That being said, a big use case would be to use the checkboxes on features that have variables that are only meaningful when the feature is active. If this PR is merged, I plan on writing a followup PR to make an **opt-in** `hint_string` value that would hide the properties in the group if the feature is not active. That change would reduce the amount of code as many of the use-cases of `_validate_property` are solely to change the editor visibility of properties. There would also probably have to be a clarification that it doesn't reset/modify any of the properties that it hides.

Repiteo reviewed (approved):
The core idea is functional & should be straightfoward to iterate upon. Let's send it as-is, giving us a healthy amount of time to evaluate and adjust during the dev cycle

Repiteo commented:
Thanks!

Raftatul commented:
Is this how it's supposed to be used in GDScript?
Unless I misunderstood something, I think it should be handled automatically.

```
@export_group("Data Group")
@export_custom(PROPERTY_HINT_GROUP_ENABLE, "") var data_group_enabled: bool :
	set(value):
		data_group_enabled = value
		notify_property_list_changed()
		
@export var content: String


func _validate_property(property: Dictionary):
	if property.name == "content" and not data_group_enabled:
		property.usage = PROPERTY_USAGE_NO_EDITOR
```

Raftatul commented:
> This PR does not change the visibility of any property other than the `enabled` property (which it moves into the section header). That being said, a big use case would be to use the checkboxes on features that have variables that are only meaningful when the feature is active. If this PR is merged, I plan on writing a followup PR to make an **opt-in** `hint_string` value that would hide the properties in the group if the feature is not active. That change would reduce the amount of code as many of the use-cases of `_validate_property` are solely to change the editor visibility of properties. There would also probably have to be a clarification that it doesn't reset/modify any of the properties that it hides.

Ah, my bad. I didn't seen that comment.

thottee commented:
All credit to the creator, but sorry, I tested it and I don't like this feature.
The biggest problem is that buttons where you can only click on a small area make it difficult to have a smooth workflow.
It's easy to click on it, miss it, and in places where you need to use it often, it's simply frustrating.
From a UX perspective, need to forget about all the possible things where you have to click tiny buttons. Especially in frequently used places.

KoBeWi commented:
See #105623
